### DISCUSSION TITLE
preprocess / coerce = coupling to view / form / controller

### DISCUSSION THREAD

**Discussion by akomm** [seasonedcc/composable-functions#124] [General]
The first example:
![grafik](https://github.com/seasonedcc/domain-functions/assets/2409779/75c25fcc-82a5-42ff-a765-3b7b2d508bdc)

Also there needs to be a separation between validation feedback on inputs that the user has control of to render in a form and those that are not affected by the user. The later failing to parse are actually application level errors. 

It might feel convenient just throwing everything in a result with error and then:
![grafik](https://github.com/seasonedcc/domain-functions/assets/2409779/246404ee-c241-4fe4-a2ef-9849528fa027)

But typically leads to more debugging what actually went wrong. Might be something wrong with the request, might be some issue querying the database, might be actually invalid param input. Might be anything.

Its better to parse actual user-controlled input and provide feedback and at least handle all other exception as an application error, potentially rendering it in an error boundary as an error - in dev provide the exception info, in prod log it.

What do you think?

**Comment by gustavoguichard** [comment]
Hey @akomm , that is a trivial example but in a real world application I'd do more checks depending on the intent.
If we are talking about a form submission we could be checking for `inputErrors`, and we can even throw inputErrors from within the domain functions:

```ts
const myDf = mdf(z.object({ email: z.string().email() }))(async ({ email }) => {
  const result = await doSomething()
  if (result.restricted && !email.endsWith(company.domain)) {
    throw new InputError("This section is restricted to employees", "email")
  }

  if (somethingElse) {
    throw new MyCustomError("Specific error")
  }
  // keep going
})
```

and later we can check for those errors in the action:
```ts
if (!result.success) {
  if (result.inputErrors.length > 0) {
    return json(result.inputErrors, 422)
  }
  if (result.errors.some(err => err.exception instanceof MyCustomError)) {
    return json({ message: "Something custom" })
  }
  throw notFound()
}
```

TLDR: the 3 kinds of errors already allow us to branch out the response but you can go further with custom errors by checking their exception ;)

**Reply by akomm** [reply]
What I mean is, `coerce` and `preprocess` looks to me like processing form submission data that is sent as string over the wire. Does not look like something you would want to have in your domain functions. 

Also one of my point was in the example that included params in the schema. So you combine form inputs with some other input source, but only part of it is something the user has influence over and can be reflected as feedback in the form.

**Reply by gustavoguichard** [reply]
Right... that example is dated but I'd probably pass the params as environment instead of input anyway.
